File: Design-Patterns/StateMachine/Cpp/StateMachine/scene_controller.cpp
#include "scene.h"

scene_controller::scene_controller()
{
	scene_ = nullptr;
}

void scene_controller::set_scene(scene* scene, bool deleteOldScene)
{
	if (scene_ != nullptr)
	{
		scene_->end();

		if (deleteOldScene)
		{
			delete scene_;
		}
	}	

	scene_ = scene;
	scene_->start();
}

void scene_controller::update() const
{
	if (scene_ != nullptr)
	{
		scene_->update();
	}
}

File: Design-Patterns/StateMachine/CSharp/StateMachine/SceneController.cs
using StateMachine.Scenes;

namespace StateMachine
{
    class SceneController
    {
        private Scene _scene;

        public SceneController()
        {
            
        }

        public void SetScene(Scene scene)
        {
            _scene?.End();
            _scene = scene;
            _scene.Start();
        }

        public void Update()
        {
            _scene?.Update();
        }
    }
}
